docs/merkletree/treetex/main.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// treetek is a command to produce LaTeX documents representing merkle trees.
+// Uses the Forest package.
+//
+// Usage: go run main.go | xelatex
+// This should generate a PDF file called treetek.pdf containing a drawing of
+// the tree.
+//
+package main
+
+import (
+	"flag"
+	"fmt"
+	"math/bits"
+	"strings"
+
+	"github.com/google/trillian/merkle"
+	"github.com/google/trillian/storage"
+)
+
+const (
+	preamble = `
+% Hash-tree
+% Author: treetek
+\documentclass[convert]{standalone}
+\usepackage[dvipsnames]{xcolor}
+\usepackage{forest}
+
+
+\begin{document}
+
+% Change colours here:
+\definecolor{inclusion}{rgb}{1,0.5,0.5}
+\definecolor{inclusion_ephemeral}{rgb}{1,0.7,0.7}
+\definecolor{perfect}{rgb}{1,0.9,0.5}
+\definecolor{target}{rgb}{0.5,0.5,0.9}
+\definecolor{target_path}{rgb}{0.7,0.7,0.9}
+
+\begin{forest}
+`
+
+	postfix = `\end{forest}
+\end{document}
+`
+	// maxLen is a suitably large maximum nodeID length for storage.NodeID.
+	maxLen = 64
+)
+
+var (
+	treeSize  = flag.Int64("tree_size", 23, "Size of tree to produce")
+	inclusion = flag.Int64("inclusion", -1, "Leaf index to show inclusion proof")
+
+	// nInfo holds nodeInfo data for the tree.
+	nInfo = make(map[string]nodeInfo)
+)
+
+// nodeInfo represents the style to be applied to a tree node.
+type nodeInfo struct {
+	incProof    bool
+	incPath     bool
+	target      bool
+	perfectRoot bool
+	ephemeral   bool
```

Comment:
Is `ephemeral == !perfectRoot` in all cases?

Reply:
Nope - `perfectRoot` is only true for the top of the perfect subtree (so we'd accidentally label internal nodes of a perfect subtree as being `ephemeral` if we did that).

Reply:
Ah, so that's compact Merkle tree nodes? Arguably, all internal nodes are also perfect, and these are kind of "topmost" out of all perfect ones.

Reply:
Yup, exactly.
I could make all perfect nodes be highlighted, but I thought that might be too cluttered.